src/java/org/apache/cassandra/index/sai/SSTableIndex.java

```diff
@@ -245,12 +246,12 @@ public void release()
 
             /*
              * When SSTable is removed, storage-attached index components will be automatically removed by LogTransaction.
-             * We only remove index components explicitly in case of index corruption or index rebuild.
+             * We only remove index components explicitly in case of index corruption or index rebuild if immutable
+             * components are not in use.
              */
-            if (obsolete.get())
-            {
-                sstableContext.indexDescriptor.deleteColumnIndex(indexContext);
-            }
+            var group = sstableContext.indexDescriptor.perIndexGroup(indexContext);
+            if (obsolete.get() && !group.version().useImmutableComponentFiles())
+                group.asWriter().forceDeleteAllComponents();
```

Comment:
If index is dropped, writer should still be able to discard index files locally..

Reply:
Looking at it again,  are they removed by `SAI#getInvalidateTask` ?

Reply:
> are they removed by `SAI#getInvalidateTask` ?

Yeah.

To be honest, I don't think the whole `SSTableIndex::markObsolete` is still useful after this patch, but I was a bit worried of breaking something unintentionally.

More precisely, my understanding is that `SSTableIndex::markObsolete` is called in 2 cases:
1. during index build (rebuild or not), by `SAI.SAIBuildingSupport::getIndexBuildTask` (or `getParallelIndexBuildTasks`).
2. during index drop, by `SAI#getInvalidateTask`.
and when we have immutable components, we want to delete in the 2nd case but not the first one, so I added deletion directly in `SAI#getInvalidateTask`.

Now, when we do not use immutable components, we still have the added deletion in `SAI#getInvalidateTask`, so that means the deletion in `SSTableIndex::markObsolete` is only maybe useful for index build. But when we build indexes for a sstable, we already delete the old files in that case (in `SAIBuilder`), so I wonder if it's truly useful. That said, I'm only so familiar with all this code and as said above, I was worried about breaking something.

But if my reasoning above reasonates, I can remove `SSTable::markObsolete` and replace it's calls by just `SSTable::release`. That might make things overall easier to follow. WDYT?

Reply:
I actually ended up keeping `SSTableIndex::markObsolete`, though I renamed it to `markIndexDropped` to hopefully make it more explicit as to when it is called, and left the deletion of the component there, but removing from `getInvalidateTask`. That's because as you pointed out in one of your other comment, we should delete stuff only after we're sure they are released, which is what the "obsolete" part was doing.

But I also replaced the call to `markObsolete` by just `release` in `IndexViewManager::drop` (which is confusingly only called before a rebuild, so I renamed it): that's case 1 above and the builder will already get rid of components as appropriate.